Remove feature gated enum variants

tcharding opened the pull request:
Feature gating enum variants makes code that uses the library brittle while we do not have `non_exhaustive`, we should avoid doing so. Instead we can add a dummy type that is available when the feature is not turned on. Doing so enables the compiler to enforce that we do not create the error type that is feature gated when the feature is not enabled.

Remove the feature gating around `bitcoincensensus` featured error enum variants.

Patch 1 is a trivial preparatory change to the `Display` impl.

Closes: #645

Kixunil commented on src/blockdata/script.rs:
```diff
@@ -122,33 +122,36 @@ pub enum Error {
+
+/// Dummy error type used when `bitcoinconsensus` feature is not enabled.
+#[cfg(not(feature = "bitcoinconsensus"))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Clone, Copy)]
```
For the backcompat trick to work you may only implement [traits that `bitcoinconsensus::Error` implements](https://docs.rs/bitcoinconsensus/latest/bitcoinconsensus/enum.Error.html#implementations)

dr-orlovsky replied:
Regarding derives on dumb error types: while we should not underderive, we can overderive. I.e., if the upstream adds derives to the respective error types this will not be a breaking change for them, but it will break our code. Thus, I suggest to do all possible derives for dumb errors, namely Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Debug (Default for errors is a bad practice, I assume).

apoelstra replied:
@dr-orlovsky I think you're wrong here because the feature gate removes our type and replaces it with `bitcoinconsensus::Error`, meaning that if we overderive, the feature-gate will *remove* derives and potentially cause compilation breakage.

dr-orlovsky replied:
This is tested in CI in this project, while underderivation may break downstream without CI in this project detecting it (when upstream adds more derives to the error type).

tcharding replied:
Ha! This discussion is mute now, the link above is to `bitcoinconsensus` v0.17.1 but we are depending on v0.19.0-3 which has all the derives on it that we want.

tcharding replied:
My bad for blindly following web rendered docs without checking the version or the code :(

tcharding replied:
Good to have you back man, hope you get fully recovered soon.

Kixunil reviewed (changes requested):
The extra traits defeat the purpose, so they definitely have to be removed. I'm still not convinced this is useful. It's not an RC fix and we will bump MSRV in the next version. I hope one of the first tasks. If you want this to be an RC fix, I'm not strongly opposed. It's not a super hard change but a bit tricky - e.g. one can shoot himself into the foot by implementing extra traits. :) I would require a careful review.

apoelstra commented:
@Kixunil did you mean to comment on a different PR? I don't see any new traits here.

apoelstra commented on src/blockdata/script.rs:
```diff
@@ -122,30 +122,34 @@ pub enum Error {
-            #[cfg(feature="bitcoinconsensus")]
-            Error::UnknownSpentOutput(ref _point) => write!(f, "unknown spent output Transaction::verify()"),
-            #[cfg(feature="bitcoinconsensus")]
+            Error::BitcoinConsensus(ref err) => write!(f, "bitcoinconsensus verification failed: {:?}", err),
```
In d9d27556eacd91923d94571d92c756ee625185ec:

I think we're not supposed to prefix errors, instead using `source`/`cause` for them to let smart error formatters do smart things with them. But I won't hold up this PR -- probably it needs to wait for the MSRV bump so that we can use `source` and do a more general error cleanup.

apoelstra reviewed (approved):
ACK d9d27556eacd91923d94571d92c756ee625185ec

Kixunil commented:
Explanation in case it's still confusing to someone else: the point of having dummy type is that whatever code does with it is also valid with actual error type when the feature is turned on to ensure additiveness of features. Thus the dummy type can only have functionality equal or less than the actual type. (Liskov principle in some sense.) In this case the traits `Copy` , `Hash`, `Ord`, `PartialOrd` are not implemented for `bitcoinconsensus::Error`.

tcharding commented:
I was just cleaning up old issues, with the comments above I think I'll just close this one. I'm not really sure what's the best approach to closing #645 once we bump MSRV, perhaps one of you could throw a comment on the issue if you have a sec. Otherwise I'll come back to it later, no doubt.

apoelstra commented:
@tcharding I think it is still worth fixing this, since the current situation is really bad. If you remove the extra `derive`d traits I'll ACK the PR as-is.

tcharding commented:
I've rebased the branch on my tree, removed the first patch, fixed the derive issue, made type alias `pub` and check the docs buid is sane. I do not have permissions to re-open the PR, are you able to do that @apoelstra please? I don't want to open a new PR and break the thread.

tcharding commented:
Might want an `rc-fix` label too?

Kixunil commented:
I believe you have to create a new one because of force push.

tcharding commented:
ok, will do. Cheers
